pkg/gatewayserver/io/io.go

```diff
@@ -86,8 +82,7 @@ type Connection struct {
 }
 
 // NewConnection instantiates a new gateway connection.
-func NewConnection(ctx context.Context, protocol string, gateway *ttnpb.Gateway, fp *frequencyplans.FrequencyPlan, scheduler *scheduling.Scheduler) *Connection {
-	ctx, cancelCtx := errorcontext.New(ctx)
+func NewConnection(ctx context.Context, cancelCtx errorcontext.CancelFunc, protocol string, gateway *ttnpb.Gateway, fp *frequencyplans.FrequencyPlan, scheduler *scheduling.Scheduler) *Connection {
```

Comment:
This looks like a dirty work-around for sharing context cancellation. Imo, the context lifetime ownership should not be shared. If `Connection` knows about cancellation (I think it should, like it was), it may expose `Context() context.Context` on which anyone can wait for done and an error. If there's an external actor that needs to disconnect (with an error), there's `Disconnect(error)` which `Connection` basically uses to `cancelCtx(err)`

cc @htdvisser please comment if you disagree

Reply:
Agreed. That was indeed a quick-fix to solve the leak in the PCN bridge. I've updated it now.